### PR TITLE
Implement a check that the time doesn't jump backwards.

### PR DESCRIPTION
**Pull request by knz**
As requested by #3087.

Question: any idea how to test this independently from fakeTime?

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5301)

### CODE REVIEW

File: util/timeutil/time.go [hunk 1]
```diff
@@ -36,20 +42,44 @@ func initFakeTime() {
...
+		}
+		if doCheck {
+			lastTime = nowFunc()
+			prevFunc := nowFunc
+			nowFunc = func() time.Time {
+				newTime := prevFunc()
+				if newTime.Before(lastTime) {
+					log.Warningf("backward time jump detected: previously %s, now %s (offset %s)", lastTime, newTime, newTime.Sub(lastTime))
+				}
+				lastTime = newTime
```
Comment (tschottdorf) [inline util/timeutil/time.go]: this needs thread safety.

Comment (tschottdorf) [inline util/timeutil/time.go]: `atomic.SwapInt64` comes to mind as handy for this.

File: util/timeutil/time_monotonic.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Raphael 'kena' Poss (knz@cockroachlabs.com)
+
+// +build clockmonotonic
```
Comment (tschottdorf) [inline util/timeutil/time_monotonic.go]: could you make this run always, at least for now? I'd like to be able to corroborate any occurrence of #3087 both in tests, acceptance and "production".

File: util/timeutil/time.go [hunk 3]
```diff
@@ -36,20 +42,44 @@ func initFakeTime() {
...
-// environment. The offset functionality is guarded by the  "clockoffset" build
-// tag - if built with that tag, the clock offset is parsed from the
-// "COCKROACH_SIMULATED_OFFSET" environment variable using time.ParseDuration,
-// which supports quasi-human values like "1h" or "1m".
+// Now returns the current local time, with optional tweaks specified by the environment:
+// - clock offsets. This is guarded by the  "clockoffset" build
+//   tag - if built with that tag, the clock offset is parsed from the
+//   "COCKROACH_SIMULATED_OFFSET" environment variable using time.ParseDuration,
+//   which supports quasi-human values like "1h" or "1m".
+// - time backwards jumps. This is guareded by the "clockmonotonic" build
```
Comment (tschottdorf) [inline util/timeutil/time.go]: s/guared/guarded/

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[util/timeutil/time.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UDlW11MiApNPgGpj:-KD-UDlW11MiApNPgGpk:321512895) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L53)):*
i see you're following the lead of `initFakeTime` here. Since you're following that lead, could you add a test?

---

*[util/timeutil/time.go, line 54 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UYr6_lavBFNcbsXT:-KD-UYr6_lavBFNcbsXU:1888310848) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L54)):*
maybe leave this on by default and turn it off only if `os.Getenv(checkMonotonicEnvKey) == "0"` - this is what we do elsewhere, see `git grep ENABLE_LOCAL_CALLS`

---

*[util/timeutil/time.go, line 80 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-ObLmljCj-6TEHUw7-r1-80:-KD-UOgb3Pwmlg1j8vz8:608616817) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L80)):*
this is also not "time backwards jumps"; it's "time monotonity checks". Importantly, it doesn't introduce jumps, which the current phrasing implies.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301)*

File: util/timeutil/time.go (hunk 1 above)
Comment (knz) [inline util/timeutil/time.go]: time.Time is larger than 64 bits unfortunately. Will use a Mutex.

Comment (tamird) [inline util/timeutil/time.go]: you can use `sync.Value`

File: util/timeutil/time_monotonic.go (hunk 2 above)
Comment (knz) [inline util/timeutil/time_monotonic.go]: Done.

File: util/timeutil/time.go (hunk 1 above)
Comment (tschottdorf) [inline util/timeutil/time.go]: `.UnixNano()` returns `int64`.

**Comment by knz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[util/timeutil/time.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UDlW11MiApNPgGpj:-KD-s-NSQoQcBtRGO8NW:-1786822450) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L53)):*
Yeah I'd like to but I'm not sure that I can do that without using your faketime code. Any suggestions? It doesn't seem right to make this test depend on faketime whereas it will be used without faketime in prod and CI.

---

*[util/timeutil/time.go, line 54 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UYr6_lavBFNcbsXT:-KD-s_IKJLS1R9oF3CO6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L54)):*
Done.

---

*[util/timeutil/time.go, line 68 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-ObLmljCj-6TEHUw7-r1-68:-KD-sdzMb5iuTMC9DB8p:-381982838) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L68)):*
Is there a problem with the current implementation?

---

*[util/timeutil/time.go, line 80 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-ObLmljCj-6TEHUw7-r1-80:-KD-ssN0Rh-nGgoZmI__:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L80)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301)*

**Comment by tamird** [comment]
Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[util/timeutil/time.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UDlW11MiApNPgGpj:-KD-u9_VLBqEyj7HUvGL:-640949724) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L53)):*
faketime is compiled in no matter what - that's why the code is arranged the way it is. Go ahead and use it.

---

*[util/timeutil/time.go, line 54 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UYr6_lavBFNcbsXT:-KD-uL0IweJXPbs8fnCI:1744993208) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L54)):*
Are you sure?

---

*[util/timeutil/time.go, line 81 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-uUMqZxvpBDktp3uY:-KD-uUMqZxvpBDktp3uZ:-1183703176) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b68e4d6e02dc648990186456fb7139c39cb72/util/timeutil/time.go#L81)):*
sorry to nit, but this is just wrong. time monotonicity checking is not a tweak to the current local time.

---

*[util/timeutil/time.go, line 88 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-uOuQmwySv9kgCqQI:-KD-uOuQmwySv9kgCqQJ:-1988423529) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b68e4d6e02dc648990186456fb7139c39cb72/util/timeutil/time.go#L88)):*
when you change the above to the `== "0"` check, please update this as well

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301)*

**Comment by knz** [comment]
Review status: 1 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[util/timeutil/time.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UDlW11MiApNPgGpj:-KD-zPPLDq7mlYOwajAc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L53)):*
Done.

---

*[util/timeutil/time.go, line 54 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UYr6_lavBFNcbsXT:-KD-zQMDdUu68DQ39idP:-775220878) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L54)):*
Yep, the environment variable is called "DISABLE" so its boolean value is inverted, as it should.

---

*[util/timeutil/time.go, line 81 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-uUMqZxvpBDktp3uY:-KD-zVZ8_ho6cL3BbP8L:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b68e4d6e02dc648990186456fb7139c39cb72/util/timeutil/time.go#L81)):*
Done.

---

*[util/timeutil/time.go, line 88 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-uOuQmwySv9kgCqQI:-KD-zWIB58L7MeyC90Ua:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87b68e4d6e02dc648990186456fb7139c39cb72/util/timeutil/time.go#L88)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[util/timeutil/time.go, line 54 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UYr6_lavBFNcbsXT:-KD01h-D8twOR6TxPeHr:-983322395) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L54)):*
why not follow the other convention? make the env key ENABLE and have "0" be the only value that turns it off.

---

*[util/timeutil/time_monotonic_test.go, line 23 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD02GdTlpgJ-_DW9U9S:-KD02GdTlpgJ-_DW9U9T:-2096265877) ([raw file](https://github.com/cockroachdb/cockroach/blob/17fbfd043945829b96d073aeb3d977979bf05da7/util/timeutil/time_monotonic_test.go#L23)):*
remove this (and now that you are using `log` in time_monotonic.go`, you can remove this import from fake_time_test as well.

---

*[util/timeutil/time_monotonic_test.go, line 27 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD02NLlZff3V-Ah34fj:-KD02NLlZff3V-Ah34fk:-1623933175) ([raw file](https://github.com/cockroachdb/cockroach/blob/17fbfd043945829b96d073aeb3d977979bf05da7/util/timeutil/time_monotonic_test.go#L27)):*
remove this newline

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[util/timeutil/time.go, line 54 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD-UYr6_lavBFNcbsXT:-KD04KyapMInzjz8F53r:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8602d0f8d7bbade62e6ca22ed8c3f67df4041101/util/timeutil/time.go#L54)):*
Done.

---

*[util/timeutil/time_monotonic_test.go, line 23 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD02GdTlpgJ-_DW9U9S:-KD03ecamg6Ym0ROgdnS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17fbfd043945829b96d073aeb3d977979bf05da7/util/timeutil/time_monotonic_test.go#L23)):*
Done.

---

*[util/timeutil/time_monotonic_test.go, line 27 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD02NLlZff3V-Ah34fj:-KD03aeHwBRaw9Pb26d_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17fbfd043945829b96d073aeb3d977979bf05da7/util/timeutil/time_monotonic_test.go#L27)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301)*

**Comment by tamird** [comment]
:lgtm: though I'd have kept time_monotonic.go and the TODO about the build tag

---

Reviewed 4 of 4 files at r5, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-:-KD05BGhkwrR16j5C8KU:2013163328)*

**Comment by knz** [comment]
@tamird: just a heads up that the benchmarks didn't find anything significant.

**Comment by tamird** [comment]
Reviewed 1 of 2 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[util/timeutil/time.go, line 40 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD0DS0L8oP208y_bFuk:-KD0DS0L8oP208y_bFul:1271484463) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46c73d69f319e66f35aeb2dcbe934e0d83c0e9c/util/timeutil/time.go#L40)):*
this change makes it impossible to set `nowFunc` back to `time.Now` (which might be desired from tests).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301)*

**Comment by knz** [comment]
Review status: 3 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[util/timeutil/time.go, line 40 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301#-KD0DS0L8oP208y_bFuk:-KD0F1OWB1qBfNcpdFEZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46c73d69f319e66f35aeb2dcbe934e0d83c0e9c/util/timeutil/time.go#L40)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5301)*

**Pull request merged**